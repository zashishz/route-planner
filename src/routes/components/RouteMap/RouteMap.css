.routeMap-container {
    width: calc(100% - 32px);
    padding: 16px;
    height: 500px;

    @include mobile {
        height: 300px;
    }
}

.routeMap-container > div {
    width: 100%;
    height: 100%;
}