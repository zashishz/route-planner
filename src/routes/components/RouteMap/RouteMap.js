import React, { Component } from 'react';
import { maps } from '../../../shared/services';

import './RouteMap.css';

class RouteMap extends Component {

    routeMapContainer = React.createRef();
    maps;
    gMap;

    init = async () => {
        this.maps = await this.props.maps();

        this.gMap = new this.maps.Map(this.routeMapContainer, {
            zoom: 7,
            center: { lat: 22.372081, lng: 114.107877 }
        })
    }

    componentDidMount() {
        this.init();
    }

    componentDidUpdate() {
        const { directions } = this.props;
        if (directions) {
            this.renderDirections(directions);
        }
    }

    interpolatePositions = path => {
        return path.map(([lat, lng]) => new this.maps.LatLng(lat, lng));
    };

    renderDirections = ({ path }) => {
        const directionsService = new this.maps.DirectionsService();
        const directionsRenderer = new this.maps.DirectionsRenderer();

        directionsRenderer.setMap(this.gMap);

        const positions = this.interpolatePositions(path);
        const waypoints = positions
            .slice(1, positions.length - 1)
            .map(location => ({ location, stopover: false }));

        const req = {
            origin: positions[0],
            destination: positions[positions.length - 1],
            optimizeWaypoints: true,
            travelMode: this.maps.TravelMode.DRIVING,
            waypoints
        };

        // get the route from directionService and then plot with the help of directionsRenderer
        directionsService.route(req, (res, flag) => {
            if (flag === this.maps.DirectionsStatus.OK) {
                directionsRenderer.setDirections(res);
            } else {
                alert('Error in direction service response');
            }
        });
    }

    render() {
        return (
            <div className="routeMap-container">
                <div ref={ (el) => this.routeMapContainer = el}></div>
            </div>
        )
    }
}

RouteMap.defaultProps = {
    maps
};

export default RouteMap;