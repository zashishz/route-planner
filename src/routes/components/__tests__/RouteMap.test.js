import React from "react";
import renderer from "react-test-renderer";

import { RouteMap } from "../RouteMap";

describe("Render RouteMap correctly", () => {
  it("render correctly", () => {
    const tree = renderer.create(<RouteMap />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
