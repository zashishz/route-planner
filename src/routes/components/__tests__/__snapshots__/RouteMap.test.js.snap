// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render RouteMap correctly render correctly 1`] = `
<div
  className="routeMap-container"
>
  <div />
</div>
`;
